### ISSUE TITLE
CTC Loss from Baidu

### ISSUE THREAD

**Issue by sbodenstein** [dmlc/mxnet#1276]
Baidu's recently released CTC loss implementation would be a nice addition to MXNet: https://github.com/baidu-research/warp-ctc

This would allow MXNet to be used for speech recognition (as mentioned in this [aspirational developers note](https://github.com/dmlc/mxnet/blob/master/doc/developer-guide/note_data_loading.md)).

**Comment by piiswrong** [comment]
since they have torch bindings, we should be able to call it directly through TorchModule. You are welcomed to submit a PR on this.

**Comment by sbodenstein** [comment]
It seems as though it doesn't have a Windows build yet anyways, so calling it as a TorchModule is probably the correct approach for now. 

Closing for now.

**Comment by xlvector** [comment]
@sbodenstein have you successfully use warp-ctc in mxnet?

**Comment by sbodenstein** [comment]
@xlvector: I have not had time to look at this again... Its on my to-do list. Have you had any luck?

**Comment by xlvector** [comment]
I find it only provide a gpu_ctc(acts, grads, labels, sizes) function to calculate grad. It's not a layer, and I do not know how to integrate them now. 

From this link https://github.com/dmlc/mxnet/blob/master/docs/how_to/torch.md, I only know how to use torch layers.

**Comment by SkidanovAlex** [comment]
@xlvector put this into a file called "mxnet_ctc.lua"

```
require('nnx')
mxnetCTC = {}
crit = nn.CTCCriterion(true):cuda()
function mxnetCTC:cuda()
    return mxnetCTC
end
function mxnetCTC:forward(data, label)
    a = crit:forward(data, torch.totable(label), torch.Tensor(data.size(data)[1]):fill(<output shape length>))
    self.output = crit.output
    return a
end
function mxnetCTC:backward(data, label)
    b = crit:backward(data, torch.totable(label))
    self.gradInput:copy(crit.gradInput)
    return b
end
return mxnetCTC
```

(replace `<output shape length>` with a single integer -- number of elements in your output sequences)
And then instantiate the layer like this:

```
mx.symbol.TorchCriterion(data=<data>, label=<label>, label_shape=(<output shape length>,), lua_string='require(\'mxnet_ctc\')')
```

The data is expected to be `(bath_size, seq_len, num_classes)` and contain the logits **before the softmax**. Label is `(batch_size, out_seq_len)`.

**Comment by ShiffnaMerza** [comment]
can i know whether using use warp-ctc in mxnet support windows

**Comment by szha** [comment]
@ShiffnaMerza yes. the source code is included in the mxnet source tree for the past few versions, and it should work for windows.
